Rays hit object and display a single color

bkach opened the pull request.

Commit:
Rays hit object and display a single color

rumblesan commented on src/Vector.hs:
```diff
@@ -13,8 +14,8 @@ sub :: Vector -> Vector -> Vector
-dotProduct :: Vector -> Vector -> Double
-(Vector x y z) `dotProduct` (Vector x1 y1 z1) = x * x1 + y * y1 + z * z1
+dot :: Vector -> Vector -> Double
+(Vector x y z) `dot` (Vector x1 y1 z1) = x * x1 + y * y1 + z * z1
```
Hah, I'd never twigged that you can define functions using infix notation like this. makes sense

bkach replied:
nice right?

rumblesan commented on src/Main.hs:
```diff
@@ -14,15 +14,143 @@
+    | otherwise = Just $ minimumBy tupleMinimum (map fromJust intersections)
+    where intersections = filter isJust $ map (intersects ray) objects
+
+tupleMinimum :: (Double, Object) -> (Double,Object) -> Ordering
```
This could do with a better name. it's a pretty specific function, but the name is very generic

bkach replied:
yep, agreed

rumblesan commented on src/Main.hs:
```diff
@@ -14,15 +14,143 @@
+    in
+        case listOfRoots of
+            [] -> Nothing
+            otherwise -> Just (minimum listOfRoots, Object (Sphere center radius) material)
```
could simplify this a touch by not recreating a new sphere
http://stackoverflow.com/questions/1153465/what-does-the-symbol-mean-in-reference-to-lists-in-haskell

bkach replied:
nice! was looking for this

rumblesan commented on src/Main.hs:
```diff
@@ -14,15 +14,143 @@
+closestIntersection ray objects
+    | null intersections = Nothing
+    | otherwise = Just $ minimumBy tupleMinimum (map fromJust intersections)
+    where intersections = filter isJust $ map (intersects ray) objects
```
so generally I use the where clauses to define functions that take args, whereas for fixed values I'll use let. This feels like it would make more sense as a let to me

bkach replied:
Here i'm just using where to allow for pattern matching. Otherwise i'd have to use an if or a case statement.

rumblesan commented on src/Main.hs:
```diff
@@ -14,15 +14,143 @@
+fovInRadians (Camera fov _ _ ) = pi * (fov / 2) / 180
+
+getPosition :: Camera -> Vector
+getPosition (Camera _ position _ ) = position
```
These three functions could easily be covered by just using record syntax for the camera and giving descriptive names for the accessor functions.

EDIT: read all three too quickly. Only the last one could be covered by record syntax. Might be worth splitting these out into a separate module though

bkach replied:
Agreed - we'll do that on another branch.

rumblesan commented on src/Main.hs:
```diff
@@ -14,15 +14,143 @@
+getColorFromIntersection (Just ( _ , Object _ (Material color))) = color
+
+-- Generating rays, assuming distance to the image is 1 unit
+generateRay :: Camera -> Int -> Int -> Int -> Int -> Ray
```
This is one of those functions that will definitely want refactoring, but only once we've got some tests for it haha

bkach replied:
yeah it's a monster!

rumblesan commented on src/Main.hs:
```diff
@@ -14,15 +14,143 @@
+minWithEmpty [] = -1
+minWithEmpty list = minimum list
+
+trace x y 
```
this should definitely have a type signature. more importantly, you can simplify it by doing the unpacking in the pattern matching. I'll chuck up a PR with an example

bkach replied:
awesome

rumblesan committed:
Can use catMaybes instead of mapping isJust

Hoogle is your friend. A quick search for [ Maybe a ] -> [ a ] found me
catMaybes ;)

rumblesan commented on src/Main.hs:
```diff
@@ -14,15 +14,143 @@
+    | otherwise = getColorFromIntersection intersection
+    where intersection = closestIntersection (generateRay camera sceneWidth sceneHeight x y) objects
+
+getColorFromIntersection (Just ( _ , Object _ (Material color))) = color
```
it seems weird that this function takes a Just. Really it should take the intersection and unwrapping the Just is handled outside of it

bkach replied:
Not sure what you mean here - the intersection is just a just?

rumblesan committed:
Tidy up trace function

* Use pattern matching for unwrapping Maybe
* getColourFromIntersection doesn't take a Maybe
* Adding type signatures

Commit:
cleaned up code

bkach commented:
You'll have to rebase your commit against the new commit i have here. Shouldn't be too much trouble!

Then we can start working on the master branch!

Commit:
Merge pull request #7 from bkach/simplifications

Simplifications

rumblesan committed:
Move colours into separate module

Also remove references to PixelRGB8 throughout main. doesn't feel good
to have a third party type crop up throughout the code

rumblesan committed:
Change Colours module name to Colour

Also qualify some of the function calls

Commit:
Merge pull request #9 from bkach/tidying-colour

Tidying colour
